webhook/main.go

```diff
@@ -0,0 +1,137 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+
+	cloudbuild "cloud.google.com/go/cloudbuild/apiv1/v2"
+	"cloud.google.com/go/cloudbuild/apiv1/v2/cloudbuildpb"
+	kms "cloud.google.com/go/kms/apiv1"
+	"github.com/abcxyz/pkg/githubauth"
+	"github.com/google/go-github/v69/github"
+	"github.com/sethvargo/go-gcpkms/pkg/gcpkms"
+	"golang.org/x/oauth2"
+)
+
+// gcloud run deploy webhook-go --region=us-west1 --source . --update-secrets=/etc/secrets/webhook/key=${KEY_NAME}:latest --allow-unauthenticated --set-env-vars=APP_ID=${APP_ID},TRIGGER_ID=${TRIGGER_ID},PROJECT_ID=${PROJECT_ID},KEY_ID=${KEY_ID}
+
+func main() {
+	log.Print("starting server")
+	http.HandleFunc("/", handler)
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "8080"
+		log.Printf("defaulting to port %s", port)
+	}
+	log.Printf("starting server on port %s", port)
+	if err := http.ListenAndServe(":"+port, nil); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func handler(resp http.ResponseWriter, req *http.Request) {
+	ctx := context.Background()
+
+	webhookSecret, err := os.ReadFile("/etc/secrets/webhook/key")
+	if err != nil {
+		log.Printf("failed to read webhook secret: %v", err)
+		resp.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	payload, err := github.ValidatePayload(req, webhookSecret)
+	if err != nil {
+		log.Printf("failed to validate payload: %v", err)
+		resp.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	event, err := github.ParseWebHook(github.WebHookType(req), payload)
+	if err != nil {
+		log.Printf("failed to parse webhook: %v", err)
+		resp.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	switch event := event.(type) {
+	case *github.WorkflowJobEvent:
+		if event.Action == nil || *event.Action != "queued" {
+			log.Printf("no action taken for %v", *event.Action)
+			resp.WriteHeader(http.StatusNoContent)
+			return
+		}
+
+		kmsClient, err := kms.NewKeyManagementClient(ctx)
+		if err != nil {
+			log.Printf("failed to create kms client: %v", err)
+			resp.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		keyId := os.Getenv("KEY_ID")
+		signer, err := gcpkms.NewSigner(ctx, kmsClient, keyId)
+		if err != nil {
+			log.Printf("failed to create app signer: %v", err)
+			resp.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		appId := os.Getenv("APP_ID")
+		app, err := githubauth.NewApp(appId, signer)
+		if err != nil {
+			log.Printf("failed to setup client: %v", err)
+			resp.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		installation, err := app.InstallationForID(ctx, strconv.FormatInt(*event.Installation.ID, 10))
+		if err != nil {
+			log.Printf("failed to setup client: %v", err)
+			resp.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		httpClient := oauth2.NewClient(ctx, installation.AllReposOAuth2TokenSource(ctx, map[string]string{
+			"administration": "write",
+		}))
+		gh := github.NewClient(httpClient)
```

Comment:
These are all expensive, one time, operations that should be done outside of the http handler.